### PR TITLE
gh-106078: Suppress the warning caused by multi-phase initialization of `decimal`

### PR DESCRIPTION
**Pull request by CharlieZhao95**
```shell
>>> python -m test test_decimal
Running Release|x64 interpreter...
0:00:00 Run tests sequentially
0:00:00 [1/1] test_decimal
...: warning: mpd_setminalloc: ignoring request to set MPD_MINALLOC a second time
```
This warnning is caused by the multi-phase initialization of `decimal` module. When `decimal` module is repeatedly imported, `decimal_exec` will be executed multiple times and this warning will be raised.

See https://github.com/python/cpython/pull/107287#discussion_r1278206860 for details.

* Issue: gh-106078

### CODE REVIEW

**Comment by CharlieZhao95** [comment]
Maybe we can skip news for this PR.

**Comment by CharlieZhao95** [comment]
It looks like the global variable check is failing, the tool cannot ignore static variables inside functions. Adding the following function might be a more elegant option.
```c
__attribute__((constructor)) void minalloc_init(void)
{
    mpd_setminalloc(_Py_DEC_MINALLOC);
}
```
The `constructor` attribute causes the function to be called automatically once at program start (before `main()`), this matches the comment for `mpd_setminalloc`.

File: Modules/_decimal/_decimal.c [hunk 1]
```diff
@@ -5898,7 +5899,12 @@ _decimal_exec(PyObject *m)
     mpd_reallocfunc = PyMem_Realloc;
     mpd_callocfunc = mpd_callocfunc_em;
     mpd_free = PyMem_Free;
-    mpd_setminalloc(_Py_DEC_MINALLOC);
+
+    /* Suppress the warning caused by multi-phase initialization */
+    if (minalloc_is_set) {
```
Comment (Eclips4) [inline Modules/_decimal/_decimal.c]: Are you sure of expression in brackets? Isn't there should been `(!minalloc_is_set)`?

Comment (CharlieZhao95) [reply Modules/_decimal/_decimal.c]: Oh thanks! I made a mistake here.
Amazingly, the tests all pass even though this code is not executed once.

**Review by Eclips4 — APPROVED** [review]
I've tested it locally, and it works. Multiple imports of `decimal` doesn't make any warnings anymore.
LGTM.

**Comment by AlexWaygood** [comment]
This warning was recently brought up again in https://github.com/python/cpython/issues/110266

**Comment by erlend-aasland** [comment]
See also the discussion about a [global init/exit API](https://discuss.python.org/t/a-new-c-api-for-extensions-that-need-runtime-global-locks/20668/13?u=erlendaasland) for extension modules.

**Review by erlend-aasland — APPROVED** [review]

**Pull request merged**